packages/snjs/lib/services/api/api_service.ts

```diff
@@ -207,34 +250,127 @@ export class SNApiService extends PureService {
-   *                    would receive parameters as params['foo'] with value equal to mfaCode.
-   * @param mfaCode     The mfa challenge response value.
-   */
-  getAccountKeyParams(
-    email: string,
-    mfaKeyPath?: string,
-    mfaCode?: string
+  private async getAuthMethods(
+    email: string
+  ): Promise<HttpResponse<GetAuthMethodsResponse>> {
+    const params = this.params({ email });
+
+    return this.request({
+      verb: HttpVerb.Get,
+      url: joinPaths(this.host, Paths.v1.authMethods),
+      fallbackErrorMessage: messages.AUTH_METHODS_ERROR,
+      params,
+    }) as Promise<HttpResponse<GetAuthMethodsResponse>>;
+  }
+
+  private async requestWithMfa(
+    httpRequest: HttpRequestWithFallbackError & {
+      identifier: string;
+    }
   ): Promise<HttpResponse> {
-    const params = this.params({
-      email: email,
-    });
-    if (mfaKeyPath && mfaCode) {
-      params[mfaKeyPath] = mfaCode;
+    const { identifier, ...restRequest } = httpRequest;
+
+    for (;;) {
+      const authMethodsResponse = await this.getAuthMethods(identifier);
+      const mfaResult = await this.mfaService.handleMfa(authMethodsResponse);
+
+      switch (mfaResult.status) {
+        case MfaStatus.ErrorMissingCode: {
+          // User dismissed window without input
+          return this.createErrorResponse(
+            messages.SignInStrings.SignInCanceledMissingMfa,
+            StatusCode.CanceledMfa
+          );
+        }
+        case MfaStatus.ErrorResponse: {
+          const originalResponse = mfaResult.response;
+          return {
+            error: {
+              message: messages.SignInStrings.ErrorFetchingAuthMethods,
+              status: originalResponse.status,
+              wrappedError: originalResponse.error,
+            },
+            status: originalResponse.status,
+          };
+        }
+        default: {
+          const mfaParams = this.mfaService.getMfaQueryParams(mfaResult);
+
+          const response = await this.request({
+            ...restRequest,
+            params: {
+              ...restRequest.params,
+              ...mfaParams,
+            },
+          });
+
+          if (response.error?.tag === ErrorTag.MfaInvalid) {
+            // alert, then iterate (this is an infinite loop)
+            await this.alertService.alert(messages.SignInStrings.IncorrectMfa);
+          } else {
+            return response;
+          }
+        }
+      }
     }
+  }
+
+  public async enableMfa({
+    userUuid,
+    secret,
+  }: {
+    userUuid: Uuid;
+    secret: string;
+  }): Promise<HttpResponse<unknown>> {
+    return this.request({
+      verb: HttpVerb.Put,
+      url: joinPaths(this.host, Paths.v1.putSetting(userUuid)),
+      fallbackErrorMessage: messages.MFA_ENABLE_ERROR,
+      authentication: this.session?.authorizationValue,
+      params: {
+        name: MfaSetting.MfaSecret,
+        value: secret,
+        serverEncryptionVersion: 1,
```

Comment:
`1` should be extracted into named const.

Reply:
- [ ] extract